core/src/main/java/org/apache/calcite/rel/rules/CombineSimpleEquivalenceRule.java

```diff
@@ -0,0 +1,226 @@
+    int spoolCounter = 0;
+
+    // Get metadata query for row count estimation
+    final RelMetadataQuery mq = call.getMetadataQuery();
+
+    // For each shared component, create a spool
+    for (RelNode sharedComponent : sharedComponents) {
+      // Get the actual row count of the shared component being materialized
+      double actualRowCount = mq.getRowCount(sharedComponent);
+
+      SpoolRelOptTable spoolTable =
+          new SpoolRelOptTable(null,  // no schema needed for temporary tables
+          sharedComponent.getRowType(),
+          "spool_" + spoolCounter++,
+          actualRowCount); // Pass the actual row count for accurate cardinality);
+
+      // Create the TableSpool that will produce/write to this table
+      LogicalTableSpool spool =
+          (LogicalTableSpool) RelFactories.DEFAULT_SPOOL_FACTORY.createTableSpool(
+              sharedComponent,
+              Spool.Type.LAZY,  // Read type
+              Spool.Type.LAZY,  // Write type
+              spoolTable);
+
+      digestToSpool.put(sharedComponent.getRelDigest(), spool);
+    }
+
+    combine =
+        combine.accept(getReplacer(digestToSpool));
+
+    call.transformTo(combine);
+  }
+
+  private static RelHomogeneousShuttle getReplacer(
+      Map<RelDigest, LogicalTableSpool> digestToSpool) {
+    Set<RelDigest> producers = new HashSet<>();
+
+    return new RelHomogeneousShuttle() {
+      @Override public RelNode visit(RelNode node) {
+        // Check if this node's digest matches any of our shared components
+        RelDigest nodeDigest = node.getRelDigest();
+        if (digestToSpool.containsKey(nodeDigest)) {
+          LogicalTableSpool spool = digestToSpool.get(nodeDigest);
+
+          if (producers.contains(nodeDigest)) {
+            // Subsequent occurrence - replace with table scan (consumer)
+            return LogicalTableScan.create(
+                node.getCluster(),
+                spool.getTable(),
+                ImmutableList.of());
+          } else {
+            // First occurrence - replace with the spool (producer)
+            producers.add(nodeDigest);
+            return spool;
+          }
+        }
+
+        return super.visit(node);
+      }
+    };
+  }
+
+
+  /** Rule configuration. */
+  @Value.Immutable(singleton = true)
+  public interface Config extends RelRule.Config {
+    Config DEFAULT = ImmutableCombineSimpleEquivalenceRule.Config.builder()
+        .build()
+        .withOperandFor(Combine.class);
+
+    @Override default CombineSimpleEquivalenceRule toRule() {
+      return new CombineSimpleEquivalenceRule(this);
+    }
+
+    default Config withOperandFor(Class<? extends Combine> combineClass) {
+      return withOperandSupplier(b -> b.operand(combineClass)
+          .predicate(combine -> {
+            // Don't fire if any immediate child is a Spool
+            for (RelNode input : combine.getInputs()) {
+              if (input instanceof Spool) {
```

Comment:
Could this situation occur? Are there only two possible cases: HepRelVertex and RelSubset?